src/wallet/rpcnames.cpp

```diff
@@ -483,11 +523,81 @@ name_firstupdate ()
   if (name.size () > MAX_NAME_LENGTH)
     throw JSONRPCError (RPC_INVALID_PARAMETER, "the name is too long");
 
-  const valtype rand = ParseHexV (request.params[1], "rand");
-  if (rand.size () > 20)
-    throw JSONRPCError (RPC_INVALID_PARAMETER, "invalid rand value");
+  const bool fixedRand = request.params.size () >= 2 && !request.params[1].isNull ();
+  const bool fixedTxid = request.params.size () >= 3 && !request.params[2].isNull ();
+  valtype rand(20);
+  if (fixedRand)
+    {
+      rand = ParseHexV (request.params[1], "rand");
+      if (rand.size () > 20)
+        throw JSONRPCError (RPC_INVALID_PARAMETER, "invalid rand value");
+    }
+
+  uint256 prevTxid = uint256::ZERO; // if it can't find a txid, force an error
+  if (fixedTxid)
+    {
+      prevTxid = ParseHashV (request.params[2], "txid");
+    }
+  else
+    {
+      // Code slightly duplicates name_scan, but not enough to be able to refactor.
+      /* Make sure the results are valid at least up to the most recent block
+         the user could have gotten from another RPC command prior to now.  */
+      pwallet->BlockUntilSyncedToCurrentChain ();
+
+      LOCK (cs_main);
+
+      for (const auto& item : pwallet->mapWallet)
+        {
+          const CWalletTx& tx = item.second;
+          if (!tx.tx->IsNamecoin ())
+            continue;
+
+          CScript output;
+          CNameScript nameOp;
+          int nOut = -1;
+          for (unsigned i = 0; i < tx.tx->vout.size (); ++i)
+            {
+              CScript curOutput = tx.tx->vout[i].scriptPubKey;
+              const CNameScript cur(curOutput);
+              if (!cur.isNameOp ())
+                continue;
+              if (cur.getNameOp () != OP_NAME_NEW)
+                continue;
+              if (nOut != -1) {
+                LogPrintf ("ERROR: wallet contains tx with multiple"
+                           " name outputs");
+                continue;
+              }
+              nameOp = cur;
+              nOut = i;
+              output = curOutput;
+            }
+
+          if (nOut == -1)
+            continue; // no name outputs found
 
-  const uint256 prevTxid = ParseHashV (request.params[2], "txid");
+          const int depth = tx.GetDepthInMainChain ();
+          if (depth <= 0)
+            continue; // not confirmed
+
+          if (!fixedRand)
+            assert (!getNameSalt (pwallet, name, output, rand)); // we should have the private key - we're sending from it!
+
+          valtype toHash(rand);
+          toHash.insert (toHash.end (), name.begin (), name.end ());
+
+          if (uint160 (nameOp.getOpHash ()) != Hash160 (toHash))
+            continue;
+
+          // found it
+          prevTxid = tx.GetHash ();
+
+          break; // if there be more than one match, the behavior is undefined
+        }
+      if (prevTxid == uint256::ZERO)
```

Comment:
I think we could move this out of the outer `if(fixedTxid)`.  It won't change behaviour, but IMHO makes the code more explicit to a reader (i.e. first we do something to determine the txid, and then we enfoce we got it in any case).